configuration - fix queue configuration default directory

albinsjolund opened the pull request:
Before:
Configuration domain.queue.default.directory was not being used.

Now:
Uses queuebase if any, else <default.directory>/<group.alias>.qb

Fixes #171

OlofGullnas commented:
I tried this branch.  Wanted to look at a few things. Built OK, but casual-make --debug test failed! Surprised me a bit as the tests went OK in the Github build.

Failed in the test casual_queue_forward.configuration_post.

Built patch/1.6/main instead (parent to this PR). It also built OK, and test suite completed without errors.
Went back to patch/1.6/ queue_configuration branch. Same test fails. When looking at the test printout it seems as if the changes applied in the "post" done in the test are missing (i.e. was not applied). The x1, x2 and x3 queues did not appear in the printout of udated.queue.

I checked casual.log and the "suspicious" thing I notice are the entries:

```
2023-03-17T12:30:22.482769+01:00|forward-domain|201deb530ccd436b95f6faee878d035d|109333|139662476830592|casual-domain-manager||||error|process cored, alias: extra, details: { pid: 109348, status: 0, reason: core}
2023-03-17T12:30:22.486170+01:00|forward-domain|201deb530ccd436b95f6faee878d035d|109333|139662476830592|casual-domain-manager||||information|process exited, alias: extra, details: { pid: 109348, status: 0, reason: core}
```
The executable that cores seems to be middleware/queue/bin/casual-queue-group. At least "sudo coredumpctl" lists a core produced by that executable...

As the "previous" commit does not have this problem, it is likely that the changes in this PR are involved. 

I tried gdb on the core file. It fails to produce a backtrace (goes into a "loop", possibly because of missing debug information) but the top of the stack displays as:

```
(gdb) bt
#0  0x00007fb49340a6d9 in ?? () from /usr/lib64/libsqlite3.so.0
#1  0x00007fb49340a865 in ?? () from /usr/lib64/libsqlite3.so.0
#2  0x00007fb49340a8f8 in ?? () from /usr/lib64/libsqlite3.so.0
#3  0x00007fb493446ad8 in ?? () from /usr/lib64/libsqlite3.so.0
#4  0x00007fb4934bd01f in ?? () from /usr/lib64/libsqlite3.so.0
#5  0x00000000004ac672 in sql::database::Connection::open (file=filesystem::path "/extra.qb" = {...})
    at ../../thirdparty/database/include/sql/database.h:649
#6  0x00000000004ac7e7 in sql::database::Connection::open (file=filesystem::path "/extra.qb" = {...})
    at ../../thirdparty/database/include/sql/database.h:664
#7  0x00000000004ac7e7 in sql::database::Connection::open (file=filesystem::path "/extra.qb" = {...})
    at ../../thirdparty/database/include/sql/database.h:664

```

A path of just "/extra.qb" is a bit suspicious :-)

OlofGullnas commented:
Much better after last commit. No longer fails the test suite. Will try to take a look at how things now work. That is if the queuebase filename handling is reasonable :-). May take a day or two (or three). Busy with building work over the weekend.

lazan commented on middleware/configuration/source/model/transform.cpp:
```diff
@@ -398,13 +398,20 @@ namespace casual
                         result.alias = group.alias.value_or( "");
                         result.note = group.note.value_or( "");
-                        result.queuebase = group.queuebase.value_or( "");
+                        result.queuebase = group.queuebase.value_or( defaults != "" ? defaults + "/" + result.alias + ".qb" : "");
```
I'm not sure that _configuration_ should be responsible for the queuebase name. At least the decision should be made in one place.
I think the _queue module_ should have the responsibility. If so, we need to extend the model.queue.group with _directory_, and make the decision when configure the queue-group with the model.

lazan commented on middleware/configuration/source/example/model.cpp:
```diff
@@ -231,7 +231,7 @@ namespace casual
       groups:
          -  alias: A
-            note: "will get default queuebase: ${CASUAL_DOMAIN_HOME}/queue/groupA.gb"
+            note: "will get default queuebase: ${CASUAL_DOMAIN_HOME}/queue/groups/A.qb"
```
Nice! You need to regenerate the documentation though. There's a _generator_ under bin that you should use. `--help` should be sufficient to understand.

albinsjolund committed:
configuration - fix queue configuration default directory

Before:
Configuration domain.queue.default.directory was not being used.

Now:
Added group.directory to internal model.
Uses queuebase if any, else if group.directory is set <group.directory>/<group.alias>.qb

Fixes #171

OlofGullnas commented:
While looking around at queue configuration related things I noticed that  domain.queue.operation.md gives the wrong default location for queuebase. It gives `${CASUAL_DOMAIN_HOM}/queue/<group-name>.qb`
as the default (the env var is incorrectly spelled...). The actual default seems to be `${CASUAL_QUEUE_DIRECTORY}/<group-name>.qb`, CASUAL_QUEUE_DIRECTORY in turn defaults to `${CASUAL_PERSISTENT_DIRECTORY}/queue/` And CASUAL_PERSISTENT_DIRECTORY defaults to `$(CASUAL_HOME_DIRECTORY}/.casual`. If only "mandatory" environment variables are defined the end result is `${CASUAL_DOMAIN_HOME}/.casual/queue/<group-name>.qb`.

I think the documented default is the "old" default location! Before is was changed to be under ".casual". There is some code in configuration::update::detail::queuebase() to move the ".qb" file from that location to the new default location. This code is only used if neither "quebase:" nor "directory:" are specified (i.e. the location of queue database files are completely defaulted). 

Should perhaps be a separate issue, but as there already are updates to queue related things in this .md file in this PR...

lazan commented on middleware/queue/source/group/handle.cpp:
```diff
@@ -586,6 +586,9 @@ namespace casual
+
+                        if( ! message.model.directory.empty())
+                           return group::Queuebase{ message.model.directory + "/" + message.model.alias + ".qb"};
                         
```
Nice!
